### PR TITLE
[KYUUBI #296] Add spark 2.4 profile and shade hive-sevice-rpc to engine jar to fix thrift build err

### PR DESCRIPTION
**Pull request by yaooqinn**
yaooqinn [PR 296](https://github.com/yaooqinn/kyuubi/pull/296) [&#10088;?&#10089;](https://pullrequestbadge.com/?utm_medium=github&utm_source=yaooqinn&utm_campaign=badge_info)

### Please add issue ID here?

Fixes #296

### Why are the changes needed?

to make new Kyuubi support 2.4.7 for backward compatibility 

### Test Plan:
- Add some test cases that check the changes thoroughly including negative and positive cases if possible
- Add screenshots for manual tests if appropriate
- [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

This PR covers the thrift part build error and let us do the rest in followups
```logtalk
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/GetCatalogs.scala:37: value catalogManager is not a member of org.apache.spark.sql.internal.SessionState
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/GetColumns.scala:180: overloaded method value listTables with alternatives:
  (db: String,pattern: String)Seq[org.apache.spark.sql.catalyst.TableIdentifier] <and>
  (db: String)Seq[org.apache.spark.sql.catalyst.TableIdentifier]
 cannot be applied to (String, String, includeLocalTempViews: Boolean)
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/GetColumns.scala:181: value getTablesByName is not a member of org.apache.spark.sql.catalyst.catalog.SessionCatalog
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/GetColumns.scala:203: value listLocalTempViews is not a member of org.apache.spark.sql.catalyst.catalog.SessionCatalog
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/GetColumns.scala:206: value schema is not a member of Any
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/GetColumns.scala:208: value table is not a member of Any
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/GetColumns.scala:208: type mismatch;
 found   : Any
 required: org.apache.spark.sql.types.StructField
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/GetColumns.scala:208: type mismatch;
 found   : Any
 required: Int
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/GetTableTypes.scala:36: value tableTypes is not a member of object org.apache.spark.sql.catalyst.catalog.CatalogTableType
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/GetTables.scala:72: overloaded method value listTables with alternatives:
  (db: String,pattern: String)Seq[org.apache.spark.sql.catalyst.TableIdentifier] <and>
  (db: String)Seq[org.apache.spark.sql.catalyst.TableIdentifier]
 cannot be applied to (String, String, includeLocalTempViews: Boolean)
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/GetTables.scala:73: value getTablesByName is not a member of org.apache.spark.sql.catalyst.catalog.SessionCatalog
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/GetTables.scala:89: overloaded method value listTables with alternatives:
  (db: String,pattern: String)Seq[org.apache.spark.sql.catalyst.TableIdentifier] <and>
  (db: String)Seq[org.apache.spark.sql.catalyst.TableIdentifier]
 cannot be applied to (String, String, includeLocalTempViews: Boolean)
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/GetTables.scala:91: value listLocalTempViews is not a member of org.apache.spark.sql.catalyst.catalog.SessionCatalog
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/schema/RowSet.scala:27: object HiveResult is not a member of package org.apache.spark.sql.execution
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/schema/RowSet.scala:108: not found: value HiveResult
[ERROR] /Users/kentyao/kyuubi/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/schema/RowSet.scala:187: not found: value HiveResult
[ERROR] 16 errors found
```

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @ulysses-you @turboFei @luzhonghao @pan3793

File: pom.xml [hunk 1]
```diff
@@ -808,5 +808,12 @@
                 <module>dev/kyuubi-tpcds</module>
             </modules>
         </profile>
+
+        <profile>
+            <id>spark-2.4</id>
+            <properties>
+                <spark.version>2.4.7</spark.version>
+            </properties>
```
Comment (turboFei) [inline pom.xml:816]: Should we also set value for hive.version here?

Comment (turboFei) [reply pom.xml:816]: https://github.com/apache/spark/blob/e0e1e21ee84bb3cb20eefee982da59afaa250a2c/pom.xml#L129-L131

**Comment by pan3793** [comment]
About the compile errors. I think we should introduce a `SparkShim` just like `HiveShim` to make it works with mutli spark versions. 
Since spark 3.1.0 was published into maven central repo accidentally, I have tested it in local, also encountered some compile errors on RowSet apis.

**Comment by pan3793** [comment]
Another things is Scala. Considering spark 2.4 support Scala 2.12 in experimental, I thinks most users use spark 2.4 with Scala 2.11, include HDP and CDH.

File: pom.xml (hunk 1 above)
Comment (yaooqinn) [reply pom.xml:816]: we dont rely on the builtin hive of spark，and it wont conflict  with that either

**Comment by yaooqinn** [comment]
> About the compile errors. I think we should introduce a `SparkShim` just like `HiveShim` to make it works with mutli spark versions.
> Since spark 3.1.0 was published into maven central repo accidentally, I have tested it in local, also encountered some compile errors on RowSet apis.

I am quite OK with this idea

**Comment by ulysses-you** [comment]
Yeah, I agree with @pan3793 . With different Spark version, Scala version will be an issue.  I believe `SparkShim` is a better idea. IMO Kyuubi has no strongly request with Scala version, isn't it @yaooqinn ?

BTW, Apache Livy also use the reflection to use Spark.

**Comment by pan3793** [comment]
> IMO Kyuubi has no strongly request with Scala version

There is unnecessary to build Kyuubi Server with multi Scala version, but kyuubi-spark-sql-engine does, because `kyuubi-spark-sql-engine` is a Spark Application. The example is Apache Zeppelin, Zeppelin Server build with Scala 2.10 & 2.11, Zeppelin Spark Interpreter build with Scala 2.10 & 2.11 & 2.12.

**Comment by yaooqinn** [comment]
> IMO Kyuubi has no strongly request with Scala version, isn't it @yaooqinn ?

The only module that cares about the scala version is the engine and should follow the spark one

**Comment by pan3793** [comment]
We also has another option, write `kyuubi-spark-sql-engine` in Java.

**Comment by yaooqinn** [comment]
the scala version does not bring compile issue with the user-side, we are not like livy which accepts scala codes.

We only need a way to be compatible with the different Spark versions that we declare to support.

Let us not make the thing too complicated and hard to maintain

**Comment by yaooqinn** [comment]
> About the compile errors. I think we should introduce a `SparkShim` just like `HiveShim` to make it works with mutli spark versions.
> Since spark 3.1.0 was published into maven central repo accidentally, I have tested it in local, also encountered some compile errors on RowSet apis.

can we open a PR for spark 3.1.0, I think we can use this Spark community's mistake to take a step ahead

**Pull request merged**